lib/utils.js

```diff
@@ -164,6 +164,68 @@ const stringify = (labels) => {
   return stableStringify(labels)
 }
 
+/**
+ *
+ * @param attrs {*[]}
+ * @returns {Object<string, string>}
+ */
+const flatOTLPAttrs = (attrs) => {
+  const flatVal = (key, val, prefix, res) => {
+    if (!val) {
+      return
+    }
+    for (const valueKey of ['stringValue', 'boolValue', 'intValue', 'doubleValue', 'bytesValue']) {
+      if (typeof val[valueKey] !== 'undefined') {
+        res[prefix + key] = `${val[valueKey]}`
+        return
+      }
+    }
+    if (val.arrayValue) {
+      val.arrayValue.values.forEach((v,i) => {
+        flatVal(`${i}`, v, `${prefix}${key}.`, res);
+      })
+      return
+    }
+    if (val.kvlistValue) {
+      flatAttrs(val.kvlistValue.values, `${prefix}${key}.`, res)
+    }
+  }
+  const flatAttrs = (attrs, prefix, res) => {
+    for (const attr of attrs) {
+      if (!attr) {
+        continue
+      }
+      flatVal(attr.key, attr.value, prefix, res)
+    }
+    return res
+  }
+  return flatAttrs(attrs, '', {})
+}
+
+/**
+ *
+ * @param attrs {Object<string, string>}
+ * @returns {{local: string, remote: string | undefined}}
+ */
+const OTLPgetServiceNames = (attrs) => {
+  let local
+  let remote
+  for (const attr of ['peer.service', 'service.name', 'faas.name', 'k8s.deployment.name', 'process.executable.name']) {
+    if (attrs[attr]) {
+      local = attrs[attr]
+      break
+    }
+  }
+  for (const attr of ['service.name', 'faas.name', 'k8s.deployment.name', 'process.executable.name']) {
+    if (attrs[attr]) {
+      remote = attrs[attr]
+      break
+    }
+  }
+  local = local || 'OTLPResourceNoServiceName'
```

Comment:
Why do we need to do this here, in my opinion service.name is the required value

Reply:
this is a port from this https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a0126a506d88df841fc4785ed7be6d41c4fb1bb9/pkg/translator/zipkin/zipkinv2/from_translator.go#L287
otlp collector does the same.

Reply:
got it.